### DISCUSSION TITLE
Steps for mutating multiple infinite lists

### DISCUSSION THREAD

**Discussion by codinginflow** [TanStack/query#7616] [Q&A] [answered]
Let's say I want to delete a resource (a post) and remove it from different feeds that use `useInfiniteQuery`

Are these the correct steps to do that?

Merely invalidating the feeds is not enough because the more pages are in cache, the longer it takes, causing quite a long delay.

I cancel the queries first because otherwise, it seems to wait for revalidation to finish before we mutate the cache.

```typescript
 const mutation = useMutation({
    mutationFn: deletePost,
    onSuccess: async () => {
      await queryClient.cancelQueries({ queryKey: ["post-feed"] });

      queryClient.setQueriesData(
        { queryKey: ["post-feed"] },
        (oldData: InfiniteData<PostsPage, unknown> | undefined) => {
          if (!oldData) return;

          return {
            ...oldData,
            pages: oldData.pages.map((page) => {
              page.posts = page.posts.filter((p) => p.id !== post.id);
              return page;
            }),
          };
        },
      );

      queryClient.invalidateQueries({ queryKey: ["post-feed"] });
      
     [...]
    },
    onError(error) {
      [...]
    },
  });
```

**Comment by TkDodo** [comment, marked as answer]
looks good

**Reply by codinginflow** [reply]
Awesome, thank you for being so responsive!

**Reply by codinginflow** [reply]
I want to remove `invalidateQueries` and keep only the cached state. This works; unless the first page hasn't been loaded yet. In this case, `cancelQueries` will keep it stuck at the initial loading state.

What is the easiest way to figure out if the first page has loaded so that I can call `invalidateQueries` conditionally? 

This is what I have right now. Is there a more efficient way?

```typescript
const results = queryClient.setQueriesData<...>(...)

const emptyQueries = results.filter((result) => !result[1]?.pages.length);      

emptyQueries.forEach(([queryKey]) => {
   queryClient.invalidateQueries({ queryKey });
});
```

**Reply by codinginflow** [reply]
OK I think invalidateQueries with a predicate is a more elegant solution:

```typescript
queryClient.invalidateQueries({
  queryKey: queryFilter.queryKey,
  predicate(query) {
    return !query.state.data;
  },
});
```
